// melange-json/Json.js
// Generated by Melange

import * as Caml_exceptions from "melange.js/caml_exceptions.js";
import * as Caml_js_exceptions from "melange.js/caml_js_exceptions.js";
import * as Caml_option from "melange.js/caml_option.js";
import * as Js__Js_exn from "melange.js/js_exn.js";

var ParseError = /* @__PURE__ */Caml_exceptions.create("Json.ParseError");

function parse(s) {
  try {
    return Caml_option.some(JSON.parse(s));
  }
  catch (exn){
    return ;
  }
}

function parseOrRaise(s) {
  try {
    return JSON.parse(s);
  }
  catch (raw_exn){
    var exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.RE_EXN_ID === Js__Js_exn.$$Error) {
      var m = exn._1.message;
      var message = m !== undefined ? m : "Unknown error";
      throw {
            RE_EXN_ID: ParseError,
            _1: message,
            Error: new Error()
          };
    }
    throw exn;
  }
}

function stringify(prim) {
  return JSON.stringify(prim);
}

export {
  ParseError ,
  parse ,
  parseOrRaise ,
  stringify ,
}
/* No side effect */
